Датасет содержит только тип активности 'IC50'

In [ ]:
import pandas as pd

# Загрузка датасета и вывод первых 5 строк
ds = pd.read_csv('dataset.csv', sep=';')
ds.head()

Molecule ChEMBL ID Molecule Name  Molecule Max Phase  Molecular Weight  \
0      CHEMBL1812417           NaN                 NaN            425.53   
1      CHEMBL1837051           NaN                 NaN            513.55   
2      CHEMBL1837047           NaN                 NaN            466.46   
3      CHEMBL1643424           NaN                 NaN            468.48   
4      CHEMBL1643427           NaN                 NaN            469.38   

   #RO5 Violations  AlogP    Compound Key  \
0              1.0   5.64     37, BSc4050   
1              1.0   4.85  22, enantiomer   
2              0.0   4.20  18, enantiomer   
3              0.0   4.84              11   
4              1.0   5.16              14   

                                              Smiles Standard Type  \
0    CCCCCCCCn1c2ccccc2c2cc(CCC(=O)O)c(OCC(=O)O)cc21          IC50   
1  O=S(=O)(c1ccc(C(F)(F)F)cc1)N1C2Cc3[nH]ncc3C1CC...          IC50   
2  O=S(=O)(c1ccc(C(F)(F)F)nc1)N1C2Cc3[nH]ncc3C1CC...          IC50   
3  COc1cc(-c2nnc3n2CCCN3C(C)c2c(F)cc(F)cc2F)ccc1-...          IC50   
4  COc1cc(-c2nnc3n2CCCN3Cc2ccc(Cl)c(Cl)c2)ccc1-n1...          IC50   

  Standard Relation  ...  Document ChEMBL ID Source ID     Source Description  \
0               '>'  ...       CHEMBL1811865         1  Scientific Literature   
1               '='  ...       CHEMBL1833811         1  Scientific Literature   
2               '='  ...       CHEMBL1833811         1  Scientific Literature   
3               '='  ...       CHEMBL1641493         1  Scientific Literature   
4               '='  ...       CHEMBL1641493         1  Scientific Literature   

       Document Journal Document Year Cell ChEMBL ID  Properties  Action Type  \
0       Bioorg Med Chem          2011  CHEMBL3308507         NaN          NaN   
1  Bioorg Med Chem Lett          2011  CHEMBL3307656         NaN          NaN   
2  Bioorg Med Chem Lett          2011  CHEMBL3307656         NaN          NaN   
3  Bioorg Med Chem Lett          2011  CHEMBL3307715         NaN          NaN   
4  Bioorg Med Chem Lett          2011  CHEMBL3307715         NaN          NaN   

   Standard Text Value  Value  
0                  NaN   80.0  
1                  NaN   36.0  
2                  NaN    4.0  
3                  NaN  342.0  
4                  NaN  126.0  

[5 rows x 48 columns]

In [ ]:
# Вывод информации по каждому столбцу датасета
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3537 entries, 0 to 3536
Data columns (total 48 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Molecule ChEMBL ID          3537 non-null   object 
 1   Molecule Name               72 non-null     object 
 2   Molecule Max Phase          47 non-null     float64
 3   Molecular Weight            3537 non-null   float64
 4   #RO5 Violations             3446 non-null   float64
 5   AlogP                       3446 non-null   float64
 6   Compound Key                3537 non-null   object 
 7   Smiles                      3537 non-null   object 
 8   Standard Type               3537 non-null   object 
 9   Standard Relation           3507 non-null   object 
 10  Standard Value              3507 non-null   float64
 11  Standard Units              3508 non-null   object 
 12  pChEMBL Value               3147 non-null   float64
 13  Data Validity Comment       13 no

Оставим только те колонки, которые отражают биологическую активность молекул:  
 7   Smiles                      3537 non-null   object  
 9   Standard Relation           3507 non-null   object  
 10  Standard Value              3507 non-null   float64  
 11  Standard Units              3508 non-null   object  

In [ ]:
# Создание нового DataFrame с выбранными столбцами и вывод первых 5 строк
df = ds[['Smiles', 'Standard Relation', 'Standard Value', 'Standard Units']]
df.head()

Smiles Standard Relation  \
0    CCCCCCCCn1c2ccccc2c2cc(CCC(=O)O)c(OCC(=O)O)cc21               '>'   
1  O=S(=O)(c1ccc(C(F)(F)F)cc1)N1C2Cc3[nH]ncc3C1CC...               '='   
2  O=S(=O)(c1ccc(C(F)(F)F)nc1)N1C2Cc3[nH]ncc3C1CC...               '='   
3  COc1cc(-c2nnc3n2CCCN3C(C)c2c(F)cc(F)cc2F)ccc1-...               '='   
4  COc1cc(-c2nnc3n2CCCN3Cc2ccc(Cl)c(Cl)c2)ccc1-n1...               '='   

   Standard Value Standard Units  
0         80000.0             nM  
1            36.0             nM  
2             4.0             nM  
3           342.0             nM  
4           126.0             nM

Выведем все единицы отношений относительно Standart Value и измерения значения Standart Value и их количество

In [12]:
df.value_counts('Standard Relation')

Standard Relation
'='     3159
'>'      345
'>='       2
'<'        1
Name: count, dtype: int64

In [13]:
print(df.value_counts('Standard Units'))

Standard Units
nM    3508
Name: count, dtype: int64


Оставим только те записи, у которых Standart Relation равен '='

In [16]:
df_filtered = df[df['Standard Relation'] == "'='"]
print(f'Количество отфильтрованных записей: {len(df_filtered)}')

Количество отфильтрованных записей: 3159


Удаление дубликатов и пропущенных значений

In [ ]:
df_filtered = df_filtered.drop_duplicates() # удаление дубликатов
df_filtered = df_filtered.dropna() # удаление пропущенных значений
print(f'Количество записей без дубликатов и пропущенных значений: {len(df_filtered)}')

Количество записей без дубликатов и пропущенных значений: 3041


In [18]:
# установка библиотеки rdkit
! pip install rdkit

   ---------------------------------------- 0.0/22.4 MB ? eta -:--:--
   ---------------------------------------- 0.3/22.4 MB ? eta -:--:--
   ----- ---------------------------------- 2.9/22.4 MB 12.0 MB/s eta 0:00:02
   ----------------- ---------------------- 9.7/22.4 MB 21.5 MB/s eta 0:00:01
   --------------------------------- ------ 18.6/22.4 MB 28.6 MB/s eta 0:00:01
   ---------------------------------------- 22.4/22.4 MB 27.3 MB/s eta 0:00:00


Проверка корретктности молекул

In [20]:
from rdkit import Chem

def is_valid_smiles(smile):
    mol = Chem.MolFromSmiles(smile)
    return mol is not None

df_filtered = df_filtered[df_filtered['Smiles'].apply(is_valid_smiles)]
print(f'Количество записей с валидными SMILE: {len(df_filtered)}')

Количество записей с валидными SMILE: 3041


Сохранение предварительно обработанного датасета

In [21]:
df_filtered.to_csv('dataset_filtered.csv', index=False)